pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java

```diff
@@ -18,225 +18,151 @@
-  public static List<File> unTar(final File inputFile, final File outputDir)
-      throws IOException, ArchiveException {
-
-    String outputDirectoryPath = outputDir.getCanonicalPath();
-    LOGGER.debug("Untaring {} to dir {}.", inputFile.getAbsolutePath(), outputDirectoryPath);
-    TarArchiveInputStream debInputStream = null;
-    InputStream is = null;
-    final List<File> untaredFiles = new LinkedList<File>();
-    try {
-      is = new GzipCompressorInputStream(new BufferedInputStream(new FileInputStream(inputFile)));
-      debInputStream = (TarArchiveInputStream) new ArchiveStreamFactory().createArchiveInputStream("tar", is);
-      TarArchiveEntry entry = null;
-      while ((entry = (TarArchiveEntry) debInputStream.getNextEntry()) != null) {
-        final File outputFile = new File(outputDir, entry.getName());
-        // Check whether the untarred file will be put outside of the target output directory.
-        if (!outputFile.getCanonicalPath().startsWith(outputDirectoryPath)) {
-          throw new IOException("Tar file must not be untarred outside of the target output directory!");
+  public static List<File> untar(File inputFile, File outputDir)
+      throws IOException {
+    String outputDirCanonicalPath = outputDir.getCanonicalPath();
+    List<File> untarredFiles = new ArrayList<>();
+    try (InputStream fileIn = Files.newInputStream(inputFile.toPath());
+        InputStream bufferedIn = new BufferedInputStream(fileIn);
+        InputStream gzipIn = new GzipCompressorInputStream(bufferedIn);
+        ArchiveInputStream tarGzIn = new TarArchiveInputStream(gzipIn)) {
+      ArchiveEntry entry;
+      while ((entry = tarGzIn.getNextEntry()) != null) {
+        String entryName = entry.getName();
+        String[] parts = StringUtils.split(entryName, ENTRY_NAME_SEPARATOR);
+        File outputFile = outputDir;
+        for (String part : parts) {
+          outputFile = new File(outputFile, part);
         }
         if (entry.isDirectory()) {
-          // creating directory
-          LOGGER.debug(String.format("Attempting to write output directory %s.", outputFile.getAbsolutePath()));
-          if (!outputFile.exists()) {
-            // does not exist
-            LOGGER.debug(String.format("Attempting to create output directory %s.", outputFile.getAbsolutePath()));
-            try {
-              // create the directory including any non-existent parent directories
-              Files.createDirectories(outputFile.toPath());
-            } catch (Exception e) {
-              LOGGER.error("Caught exception while creating directory: {}, error: {}", outputFile.getAbsolutePath(), e);
-              throw e;
-            }
-          } else {
-            // directory already exists
-            LOGGER.error("The directory already there. Deleting - " + outputFile.getAbsolutePath());
-            FileUtils.deleteDirectory(outputFile);
+          if (!outputFile.getCanonicalPath().startsWith(outputDirCanonicalPath)) {
```

Comment:
When will this not be true? `File outputFile = outputDir` should ensure it is always true?

Reply:
There is a test case for this. We want to prevent malicious tar file to inject file outside of the outputDir, and we don't know if there is some way to directly jump out of the outputDir, so this is the safest way. This security check was added in #2835